metabase/metabase
# Session activity timeout

[Product doc](https://www.notion.so/metabase/Offer-more-control-over-session-timeouts-bc8e84e49de44de0950385fc839b8df3)
Fixes https://github.com/metabase/metabase-enterprise/issues/326

## Changes

This PR adds the new session timeout setting. It defines a timespan after which users should be logged out. This means all open tabs will be redirected to the login screen automatically on timeout.

## Technical details

Session timeout functionality is added as middleware on the backend and the frontend. 

The backend middleware adds a new cookie `metabase.TIMEOUT` on login, which is a non http-only clone of the `metabase.SESSION` cookie. On every request following login, the `expires` attribute of these two cookies will be updated according to the session-timeout setting if the cookies haven't expired by the time of the request.

On the frontend, every open tab will independently and repeatedly check the presence of the `metabase.TIMEOUT` cookie, which will log the user out when it expires.

## How to verify
- Run Metabase EE
- Open a couple of tabs with random questions and dashboards and keep them in background
- In a new tab open Admin Settings -> [Authentication](http://localhost:3000/admin/settings/authentication)
- Set the session timeout to one minute
- Wait for one minutes without doing any new requests to the backend
- Ensure all the tabs show the login screen after the timeout
- Login back in any of the tabs
- Ensure all tabs are logged in and show the previous url you had before the logout

TODO: 
- [x] Add setting for timeout duration in minutes on BE
- [x] Set this setting from the FE
- [x] Log out from all active tabs on the FE after the timeout
